tslearn/early_classification.py

```diff
@@ -0,0 +1,312 @@
+
+        for idx, current_classe in enumerate(y_classes):
+            y[y == current_classe] = y_classes_indices[idx]
+
+        self.cluster_ = TimeSeriesKMeans(n_clusters=self.n_clusters)
+        self.classifiers_ = {
+            t: clone(self.base_classifier)
+            for t in range(self.minimum_time_stamp, X.shape[1] + 1)
+        }
+        self.pyhatyck_ = {}
+        self.indice_ck_ = []
+        self.__n_classes_ = len(y_classes_indices)
+        self.__len_X_ = X.shape[1]
+        c_k = self.cluster_.fit_predict((to_time_series_dataset(X)))
+        X_and_cluster = np.concatenate((X, c_k[:, np.newaxis]), axis=1)
+        X1, X2, y1, y2 = train_test_split(X_and_cluster, y, test_size=0.5)
+        X1 = X1[:, :-1]
+        c_k2 = X2[:, -1]
+        X2 = X2[:, :-1]
+        vector_of_ones = np.ones((len(X[:]),))
+        self.pyck_ = coo_matrix(
+            (vector_of_ones, (y, c_k)),
+            shape=(self.__n_classes_, self.n_clusters),
+        ).toarray()
+        for k in range(0, self.n_clusters):
+            self.pyhatyck_["pyhatycks{0}".format(k)] = []
+            self.indice_ck_.append(np.where(c_k == k))
+            current_sum = self.pyck_[:, k].sum()
+            for current_classe in range(0, self.__n_classes_):
+                self.pyck_[current_classe, k] = (
+                    self.pyck_[current_classe, k] / current_sum
+                )
+
+        for t in range(self.minimum_time_stamp, X.shape[1] + 1):
+            self.classifiers_[t].fit(X1[:, :t], y1)
+            for k in range(0, self.n_clusters):
+                index_cluster = np.where(c_k2 == k)
+
+                if len(index_cluster[0]) != 0:
+                    X2_current_cluster = np.squeeze(
+                        X2[index_cluster, :t], axis=0
+                    )
+                    y2_current_cluster = y2[tuple(index_cluster)]
+                    y2_hat = self.classifiers_[t].predict(
+                        X2_current_cluster[:, :t]
+                    )
+                    conf_matrix = confusion_matrix(
+                        y2_current_cluster, y2_hat, labels=y_classes_indices, normalize="true"
+                    )
+                    np.fill_diagonal(conf_matrix, 0)
+                    self.pyhatyck_["pyhatycks{0}".format(k)].append(
+                        conf_matrix
+                    )
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=2)
+        distances_clusters = np.asarray(distances_clusters)
```

Comment:
Previous line already creates a numpy array, so that this line is a useless copy.